### ISSUE TITLE
Hi！I have a problem about aggregated face. 

### ISSUE THREAD

**Issue by wqz960** [D-X-Y/landmark-detection#26]
Thank you for your codes, And I am confused about the aggregated face producing，How does cycleGAN learn unsupervisedly through aggregated features? Is K set arbitrarily or is it equal to 3

**Comment by D-X-Y** [comment]
We cluster faces into different clusters in an unsupervised way. And the cycleGAN learns to transfer the style from one cluster to another cluster. The K is set to be 3 as it gives us the best performance on the validation set.

**Comment by wqz960** [comment]
Thank you for you quick respond!
In your paper , in my understand, the input of cycleGAN is the features from the ResNet, but the output of CycleGAN and groundtrue of CycleGAN is what? can you tell me more detail information, I am very interested in you network!

**Comment by D-X-Y** [comment]
The input of cycleGAN is the original image. The ResNet features are only used for clustering features into K hidden styles.

**Comment by wqz960** [comment]
the original images are divided into k parts by K-means? But in CycleGAN, is it trained in unsupervised way? the output of cycleGAN should be what? Thank you!

**Comment by D-X-Y** [comment]
Taking K=3 as an example. We first train ResNet based on images with hand-crafted designed 4 styles (300W-Style). Then, we cluster the original images into K=3 parts by k-means with learned ResNet features.
Lastly, we train one cycleGAN for every two clusters.
The output of one cycleGAN would be expected to output faces with its corresponding styles, such as gray or others (we cannot accurately define a style).
The aggregated face is the average of multiple cycleGAN outputs.

**Comment by wqz960** [comment]
where are the target image of each CycleGAN output from? are they hand-crafted? because of the bad network, i still cannot reach the dataset now, so I cannot try the codes! In my understanding, the cluster style cannot be defined as a accuracy style, so what are the target images of a cluster style? Thank you! @D-X-Y

**Comment by D-X-Y** [comment]
We think the cluster styles give some implicit hidden styles. If you see Fig.6 of our paper, you can recognize the mean face from different clusters is different from each other. We think that represents a style although we can not give a definition of this "unsupervised-find" style.
Suppose we cluster images into A/B/C three clusters. We train cycleGAN on A&B, A&C, B&C.

**Comment by wqz960** [comment]
yes! But what confused me is how to train the cycleGAN, I am not good at watching codes, too many... if i feed the original image from cluster A , from what you said, the output of cycleGAN should be the corresponding images in cluster B that are transformed from cluster A , but how to get the ground true corresponding images in cluster B? And if there are three clusters, A,B,C, the image is belong to A, two GANs will perform A->B, A->C, but another image in cluster B, is also transformed to A and C on the same network?  if I am wrong in the process, please correct me. Really thank you for your patience and professional responds!

**Comment by D-X-Y** [comment]
Training cycleGAN does not require paired ground truth, please see more details in https://arxiv.org/abs/1703.10593 .
In our experiments, we only train cycleGAN on two (largest) clusters for efficiency.
Suppose these two are A & B, after training there will be two generative models A->B and B->A.
When a testing image (I) comes, we perform both model-1(A->B) and model-2(B->A) on the image,
the aggregated image is (model-1(I) + model-2(I)) / 2

**Comment by wqz960** [comment]
if there is a image in cluster C, in the process of producing aggregated face, the image is feed to the generator(A->B) and generator(B->A)? in the process of training, only using the images in cluster A and cluster B for training? Thank you!
